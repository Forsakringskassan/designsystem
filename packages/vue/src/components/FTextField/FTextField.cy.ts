import { defineComponent } from "vue";
import {
    DensityWrapper,
    SizeWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
    sizeWrapperHeight,
    sizeWrapperWidth,
} from "@fkui/test-utils/vue";
import { FIcon } from "../FIcon";
import { FTextFieldPageObject } from "../../cypress";
import FTextFieldGridExample from "./examples/FTextFieldGrid.vue";
import FTextField from "./FTextField.vue";

const ValidationComponent = defineComponent({
    template: /* HTML */ `
        <f-text-field
            v-model="field"
            v-validation.required.maxLength="{ maxLength: { length: 10 } }"
        >
            label
        </f-text-field>
    `,
    components: {
        FTextField,
    },
    data() {
        return {
            field: "",
        };
    },
});

const AppendInnerContentComponent = defineComponent({
    template: /* HTML */ `
        <f-text-field v-model="field" v-validation.required>
            label
            <template #append-inner> kr </template>
        </f-text-field>
    `,
    components: {
        FTextField,
    },
    data() {
        return {
            field: "",
        };
    },
});

const SizesComponent = defineComponent({
    template: /* HTML */ `
        <size-wrapper>
            <f-text-field v-model="field"> label </f-text-field>
        </size-wrapper>
    `,
    components: {
        SizeWrapper,
        FTextField,
    },
    data() {
        return {
            field: "",
        };
    },
});

const SizesButtonsComponent = defineComponent({
    template: /* HTML */ `
        <size-wrapper>
            <f-text-field v-model="field">
                label
                <template #input-left>
                    <button
                        type="button"
                        aria-label="left"
                        :style="buttonStyle"
                    >
                        <f-icon name="calendar" :style="iconStyle"></f-icon>
                    </button>
                </template>
                <template #input-right>
                    <button
                        type="button"
                        aria-label="right"
                        :style="buttonStyle"
                    >
                        <f-icon name="calendar" :style="iconStyle"></f-icon>
                    </button>
                </template>
            </f-text-field>
        </size-wrapper>
    `,
    components: {
        SizeWrapper,
        FTextField,
        FIcon,
    },
    data() {
        return {
            field: "",
            buttonStyle:
                "background: transparent; border: 0; flex: 0 0 auto; padding: 0;",
            iconStyle:
                "color: hotpink; width: 2rem; height: 2rem; verticalAlign: middle;",
        };
    },
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
it.skip(`should contain input field with label`, () => {
    cy.viewport(sizeWrapperWidth, sizeWrapperHeight);
    cy.mount(SizesComponent);
    cy.toMatchScreenshot();
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
it.skip(`should display icons before and after input field`, () => {
    cy.viewport(sizeWrapperWidth, sizeWrapperHeight);
    cy.mount(SizesButtonsComponent);
    cy.toMatchScreenshot();
});

describe("required validation", () => {
    const textField = new FTextFieldPageObject(".text-field");

    beforeEach(() => {
        cy.viewport(640, 200);
    });

    // afterEach(() => {
    //     cy.toMatchScreenshot();
    // });

    it(`valid`, () => {
        cy.mount(ValidationComponent);
        textField.input().type("foo").blur();
    });

    it(`invalid`, () => {
        cy.mount(ValidationComponent);
        textField.input().type("Exceeds maxlength").focus().blur();
    });
});

it("should expose error label with hidden text when emitting componentValidity event", () => {
    const onComponentValidity = cy.spy().as("onComponentValidity");
    cy.mount(
        {
            template: /* HTML */ `
                <div @component-validity="onComponentValidity">
                    <f-text-field v-validation.required>
                        <span class="sr-only"> lorem ipsum </span>
                    </f-text-field>
                </div>
            `,
            components: { FTextField },
            emits: ["componentValidity"],
            methods: {
                onComponentValidity(event: CustomEvent) {
                    this.$emit("componentValidity", event.detail);
                },
            },
        },
        {
            props: {
                onComponentValidity,
            },
        },
    );
    cy.get("input").focus();
    cy.get("input").blur();
    cy.get("@onComponentValidity").should("have.been.calledWithMatch", {
        isValid: false,
        errorMessage: "lorem ipsum",
    });
});

describe("FTextFieldGrid example", () => {
    it("should set grid classes", () => {
        cy.mount(FTextFieldGridExample);
        cy.get(".i-width-md-9 label").should("exist");
        cy.get(".i-width-md-6 input").should("exist");
    });
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
describe.skip("Append inner content", () => {
    it(`should show content`, () => {
        cy.mount(AppendInnerContentComponent);
        cy.get(".text-field__append-inner").should("contain.text", "kr");
        cy.toMatchScreenshot();
    });

    it(`should handle input padding for long input-text`, () => {
        cy.mount(AppendInnerContentComponent);
        const textField = new FTextFieldPageObject(".text-field");
        textField.enter(
            "a long text to see if the padding for the append inner content works as we will",
        );
        cy.toMatchScreenshot();
    });
});

/* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
describe.skip("density", () => {
    it(`should be densified`, () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-text-field>
                        <template #default> Inmatningsfält </template>
                        <template #append-inner>\\o/</template>
                    </f-text-field>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FTextField,
            },
        });

        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });

    it(`inline should be densified`, () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-text-field inline> Inmatningsfält </f-text-field>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FTextField,
            },
        });

        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });
});

describe("toggle existence", () => {
    it("should retain correct error message", () => {
        // Given
        const textField = new FTextFieldPageObject(`[data-test="textField"]`);
        const toggleButtonSelector = `[data-test="toggleButton"]`;

        const TestComponent = defineComponent({
            template: /* HTML */ `
                <button
                    data-test="toggleButton"
                    type="button"
                    class="button"
                    @click="onClickToggleShowTextField"
                >
                    Toggle showing text field
                </button>
                <f-text-field
                    v-if="show"
                    v-model="value"
                    v-test="'textField'"
                    v-validation.required.minLength.maxLength="{
                minLength: { length: 3 },
                maxLength: { length: 5 }
            }"
                >
                    Test
                </f-text-field>
            `,
            components: { FTextField },
            data() {
                return {
                    show: true,
                    value: "",
                };
            },
            methods: {
                onClickToggleShowTextField(): void {
                    this.show = !this.show;
                },
            },
        });

        cy.mount(TestComponent);

        textField.enter("Too many characters");
        textField.label.errorMessage().should("be.visible");
        textField.label
            .errorMessage()
            .invoke("text")
            .invoke("trim")
            .then((expectedErrorMessage) => {
                // When
                cy.get(toggleButtonSelector).click();
                textField.el().should("not.exist");
                cy.get(toggleButtonSelector).click();
                textField.el().should("be.visible");

                // Then
                textField.label.errorMessage().should("be.visible");
                textField.label
                    .errorMessage()
                    .invoke("text")
                    .invoke("trim")
                    .should("equal", expectedErrorMessage);
            });
    });
});
