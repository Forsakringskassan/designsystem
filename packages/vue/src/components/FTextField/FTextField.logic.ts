export function resolveWidthClass(
    words: string,
    inline: boolean,
): string | undefined {
    return inline
        ? undefined
        : words
              .split(" ")
              .map((word) => `i-width-${word}`)
              .join(" ");
}
