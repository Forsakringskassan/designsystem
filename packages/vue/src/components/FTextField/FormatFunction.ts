/**
 * @public
 */
export type FormatFunction<TModel> = (modelValue: TModel) => string | undefined;
