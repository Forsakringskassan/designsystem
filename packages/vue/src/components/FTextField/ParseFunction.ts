/**
 * @public
 */
export type ParseFunction<TModel> = (viewValue: string) => TModel | undefined;
