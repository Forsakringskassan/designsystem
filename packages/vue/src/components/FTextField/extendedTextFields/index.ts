export { FEmailTextField } from "./FEmailTextField";
export { FPhoneTextField } from "./FPhoneTextField";
export { FCurrencyTextField } from "./FCurrencyTextField";
export { FSearchTextField } from "./FSearchTextField";
export { default as FBankAccountNumberTextField } from "./FBankAccountNumberTextField/FBankAccountNumberTextField.vue";
export { default as FBankgiroTextField } from "./FBankgiroTextField/FBankgiroTextField.vue";
export { default as FClearingnumberTextField } from "./FClearingnumberTextField/FClearingnumberTextField.vue";
export { default as FNumericTextField } from "./FNumericTextField/FNumericTextField.vue";
export { default as FPersonnummerTextField } from "./FPersonnummerTextField/FPersonnummerTextField.vue";
export { default as FPlusgiroTextField } from "./FPlusgiroTextField/FPlusgiroTextField.vue";
export { default as FPostalCodeTextField } from "./FPostalCodeTextField/FPostalCodeTextField.vue";
export { default as FPercentTextField } from "./FPercentTextField/FPercentTextField.vue";
export { default as FOrganisationsnummerTextField } from "./FOrganisationsnummerTextField/FOrganisationsnummerTextField.vue";
