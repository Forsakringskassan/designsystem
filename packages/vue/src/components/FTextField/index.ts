export { default as FTextField } from "./FTextField.vue";
export * from "./extendedTextFields";
export { type FormatFunction } from "./FormatFunction";
export { type ParseFunction } from "./ParseFunction";
export {
    useTextFieldSetup,
    type TextFieldSetupProps,
} from "./useTextFieldSetup";
