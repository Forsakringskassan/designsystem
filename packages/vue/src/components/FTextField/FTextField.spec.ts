import "html-validate/jest";
import {
    type PendingValidityEvent,
    type ValidatableHTMLElement,
    ValidationService,
    type ValidatorConfigs,
    type ValidityEvent,
} from "@fkui/logic";
import flushPromises from "flush-promises";
import { mount, VueWrapper } from "@vue/test-utils";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import FTextField from "./FTextField.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
    options = {},
} = {}): VueWrapper<InstanceType<typeof FTextField>> {
    return mount(FTextField, {
        attrs: { ...attrs },
        props: { ...props },
        slots: { default: "Label", ...slots },
        attachTo: createPlaceholderInDocument(),
        ...options,
        global: {
            stubs: ["f-icon"],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot with label and input", () => {
        const wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with label, error message and input", () => {
        const wrapper = createWrapper({
            slots: { "error-message": "ERROR_MESSAGE" },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with label, tooltip, description, error message and input", () => {
        const wrapper = createWrapper({
            slots: {
                description: "DESCRIPTION",
                tooltip: "TOOLTIP",
                "error-message": "ERROR_MESSAGE",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it.each`
        validityMode | isValid
        ${"VALID"}   | ${true}
        ${"ERROR"}   | ${false}
        ${"INITIAL"} | ${true}
        ${"INITIAL"} | ${false}
    `(
        "should match snapshot when validityMode is $validityMode and isValid is $isValid",
        async ({ validityMode, isValid }) => {
            const wrapper = createWrapper({
                attrs: { id: "elementId" },
            });

            const input = wrapper.get("input");
            const htmlInput = input.element as HTMLInputElement;

            htmlInput.dispatchEvent(
                new CustomEvent<ValidityEvent>("validity", {
                    detail: {
                        target: wrapper.element as ValidatableHTMLElement,
                        elementId: "elementId",
                        isValid,
                        validityMode,
                        validationMessage: "Something went wrong.",
                        nativeEvent: "input",
                    },
                }),
            );
            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.element).toMatchSnapshot();
        },
    );

    it.each`
        slotsSpecified
        ${"specified"}
        ${"not specified"}
    `(
        "should match snapshot when label and description slots are $slotsSpecified",
        async ({ slotsSpecified }) => {
            const wrapper = mount(FTextField, {
                data() {
                    return {
                        defaultText: "My defaultText",
                        descriptionText: "My descriptionText",
                        descriptionScreenReaderText:
                            "My descriptionScreenReaderText",
                        discreteDescriptionText: "My discreteDescriptionText",
                        discreteDescriptionScreenReaderText:
                            "My discreteDescriptionScreenReaderText",
                    };
                },
                attachTo: createPlaceholderInDocument(),
                global: {
                    stubs: ["f-icon"],
                },
                slots:
                    slotsSpecified === "specified"
                        ? {
                              default: "Custom default slot",
                              description: "Custom description slot",
                          }
                        : {},
            });

            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.element).toMatchSnapshot();
        },
    );
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: true,
                required: true,
            },
        });
        const input = wrapper.get("input");

        expect(input.attributes("disabled")).toBeDefined();
        expect(input.attributes("required")).toBeDefined();
    });

    it("should set type", () => {
        const wrapper = createWrapper({
            props: {
                type: "email",
            },
        });
        const input = wrapper.get("input");
        expect(input.attributes("type")).toBe("email");
    });

    it("should set type to text as default", () => {
        const wrapper = createWrapper();
        const input = wrapper.get("input");
        expect(input.attributes("type")).toBe("text");
    });

    describe("inline", () => {
        it("should not set class by default", () => {
            const wrapper = createWrapper();
            expect(wrapper.classes()).not.toContain("text-field--inline");
        });

        it("should set class when enabled", () => {
            const wrapper = createWrapper({
                props: {
                    inline: true,
                },
            });
            expect(wrapper.classes()).toContain("text-field--inline");
        });
    });
});

describe("events", () => {
    it("should emit model update event when no validation is used", async () => {
        const wrapper = createWrapper();
        const input = wrapper.get("input");
        input.setValue("foo");
        await input.trigger("change");

        expect(wrapper.emitted("update")![0][0]).toBe("foo");
    });

    it("should pass listeners", async () => {
        const focus = jest.fn();
        const change = jest.fn();
        const blur = jest.fn();

        const wrapper = createWrapper({
            attrs: {
                onFocus: focus,
                onChange: change,
                onBlur: blur,
            },
        });
        const input = wrapper.get("input");
        await input.trigger("focus");
        await input.trigger("change");
        await input.trigger("blur");

        expect(focus).toHaveBeenCalled();
        expect(change).toHaveBeenCalled();
        expect(blur).toHaveBeenCalled();
    });

    it('should have ValidityMode INITIAL when "pending-validity" event is triggered', async () => {
        const wrapper = createWrapper({
            attrs: { id: "elementId" },
        });

        const input = wrapper.get("input");
        const htmlInput = input.element as HTMLInputElement;

        htmlInput.dispatchEvent(
            new CustomEvent<ValidityEvent>("validity", {
                detail: {
                    target: wrapper.element as ValidatableHTMLElement,
                    elementId: "elementId",
                    isValid: true,
                    validityMode: "ERROR",
                    validationMessage: "Something went wrong.",
                    nativeEvent: "input",
                },
            }),
        );
        await flushPromises();
        wrapper.vm.$forceUpdate();

        expect(wrapper.vm.$data.validityMode).toBe("ERROR");

        htmlInput.dispatchEvent(
            new CustomEvent<PendingValidityEvent>("pending-validity"),
        );
        await flushPromises();
        wrapper.vm.$forceUpdate();

        expect(wrapper.vm.$data.validityMode).toBe("INITIAL");
    });
});

describe("validation", () => {
    it("should display correct error message when multiple validators", async () => {
        const wrapper = createWrapper({
            options: { sync: false },
        });
        await flushPromises();

        const input = wrapper.get("input");
        const validatorConfigs: ValidatorConfigs = { number: {}, integer: {} };
        ValidationService.addValidatorsToElement(
            input.element as HTMLInputElement,
            validatorConfigs,
        );

        await input.trigger("focus");
        input.setValue("5.5a");
        await input.trigger("blur");

        expect(wrapper.get(".label__message--error").text()).toBe(
            "Du har fyllt i ett ogiltigt tecken. Fyll i siffror.",
        );

        await input.trigger("focus");
        input.setValue("5.5");
        await input.trigger("blur");

        expect(wrapper.get(".label__message--error").text()).toBe(
            "Fyll i siffror utan decimal.",
        );
    });
});

describe("formatting and parsing combined with validation", () => {
    it.each`
        valid    | nativeEvent | expected
        ${true}  | ${"input"}  | ${"not trigger"}
        ${false} | ${"change"} | ${"trigger"}
        ${true}  | ${"change"} | ${"trigger"}
        ${false} | ${"blur"}   | ${"trigger"}
        ${true}  | ${"blur"}   | ${"trigger"}
        ${false} | ${"input"}  | ${"not trigger"}
    `(
        'should $expected update event when valid="$valid" and nativeEvent="$nativeEvent"',
        async ({ valid, nativeEvent, expected }) => {
            const wrapper = createWrapper({
                attrs: {
                    id: "elementId",
                    "data-validation": true,
                },
            });

            const input = wrapper.get("input");
            const htmlInput = input.element as HTMLInputElement;
            input.setValue("qweRTY");

            htmlInput.dispatchEvent(
                new CustomEvent<ValidityEvent>("validity", {
                    detail: {
                        target: wrapper.element as ValidatableHTMLElement,
                        elementId: "elementId",
                        isValid: valid,
                        validityMode: valid ? "VALID" : "ERROR",
                        validationMessage: "Some validationmessage",
                        nativeEvent: nativeEvent,
                    },
                }),
            );

            await flushPromises();
            wrapper.vm.$forceUpdate();

            if (expected === "trigger") {
                /* eslint-disable-next-line jest/no-conditional-expect -- technical debt */
                expect(wrapper.emitted().update).toBeTruthy();
            } else {
                /* eslint-disable-next-line jest/no-conditional-expect -- technical debt */
                expect(wrapper.emitted().update).toBeFalsy();
            }
        },
    );

    it.each`
        initialModel | inputValue  | formatter | parser   | expectedModel | expectedValue
        ${""}        | ${"qweRTY"} | ${"yes"}  | ${"no"}  | ${"QWERTY"}   | ${"QWERTY"}
        ${""}        | ${"qweRTY"} | ${"no"}   | ${"yes"} | ${"qwerty"}   | ${"qweRTY"}
        ${""}        | ${"qweRTY"} | ${"yes"}  | ${"yes"} | ${"qwerty"}   | ${"QWERTY"}
        ${""}        | ${"qweRTY"} | ${"no"}   | ${"no"}  | ${"qweRTY"}   | ${"qweRTY"}
        ${"qwerty"}  | ${"qweRTY"} | ${"yes"}  | ${"yes"} | ${"qwerty"}   | ${"QWERTY"}
        ${"qwerty"}  | ${"qweRTY"} | ${"no"}   | ${"yes"} | ${"qwerty"}   | ${"qweRTY"}
    `(
        'should have model="$expectedModel", value="$expectedValue" on blur with input "$inputValue" (initialModel="$initialModel", formatter=$formatter, parser=$parser)',
        async ({
            initialModel,
            inputValue,
            formatter,
            parser,
            expectedModel,
            expectedValue,
        }) => {
            const formatterMock = (viewValue: string): string =>
                viewValue.toUpperCase();
            const parserMock = (viewValue: string): string =>
                viewValue.toLowerCase();

            const wrapper = createWrapper({
                props: {
                    formatter: formatter === "yes" ? formatterMock : undefined,
                    parser: parser === "yes" ? parserMock : undefined,
                    modelValue: initialModel,
                },
                attrs: {
                    id: "elementId",
                    "data-validation": true,
                },
            });

            const input = wrapper.get("input");
            const htmlInput = input.element as HTMLInputElement;
            input.setValue(inputValue);

            htmlInput.dispatchEvent(
                new CustomEvent<ValidityEvent>("validity", {
                    detail: {
                        target: wrapper.element as ValidatableHTMLElement,
                        elementId: "elementId",
                        isValid: true,
                        validityMode: "VALID",
                        validationMessage: "Some validationmessage",
                        nativeEvent: "change",
                    },
                }),
            );

            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.emitted("update")![0][0]).toEqual(expectedModel);
            expect(htmlInput.value).toEqual(expectedValue);
        },
    );

    it.each`
        initialModel | inputValue  | formatted | parsed       | expectedModel | expectedValue
        ${""}        | ${"qweRTY"} | ${"zero"} | ${0}         | ${0}          | ${"zero"}
        ${"qwerty"}  | ${"qweRTY"} | ${"foo"}  | ${undefined} | ${"qweRTY"}   | ${"foo"}
    `(
        'should have model="$expectedModel", value="$expectedValue" on blur with input "$inputValue" (initialModel="$initialModel", formatted=$formatted, parsed=$parsed)',
        async ({
            initialModel,
            inputValue,
            formatted,
            parsed,
            expectedModel,
            expectedValue,
        }) => {
            const formatterMock = (): string => formatted;
            const parserMock = (): number | undefined => parsed;

            const wrapper = createWrapper({
                props: {
                    formatter: formatterMock,
                    parser: parserMock,
                    modelValue: initialModel,
                },
                attrs: {
                    id: "elementId",
                    "data-validation": true,
                },
            });

            const input = wrapper.get("input");
            const htmlInput = input.element as HTMLInputElement;
            input.setValue(inputValue);

            htmlInput.dispatchEvent(
                new CustomEvent<ValidityEvent>("validity", {
                    detail: {
                        target: wrapper.element as ValidatableHTMLElement,
                        elementId: "elementId",
                        isValid: true,
                        validityMode: "VALID",
                        validationMessage: "Some validationmessage",
                        nativeEvent: "change",
                    },
                }),
            );

            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.emitted("update")![0][0]).toEqual(expectedModel);
            expect(htmlInput.value).toEqual(expectedValue);
        },
    );
});

describe("set v-model programmatic", () => {
    it.each`
        vModel
        ${""}
        ${"an initial value"}
    `(
        "should set viewModel to '$vModel' when setting v-model to '$vModel'",
        async ({ vModel }) => {
            const wrapper = createWrapper({
                attrs: { id: "elementId" },
                props: {
                    modelValue: "original input",
                },
            });

            const input = wrapper.get("input");
            input.setValue("original input");

            const htmlInput = input.element as HTMLInputElement;

            htmlInput.dispatchEvent(
                new CustomEvent<ValidityEvent>("validity", {
                    detail: {
                        target: wrapper.element as ValidatableHTMLElement,
                        elementId: "elementId",
                        isValid: true,
                        validityMode: "ERROR",
                        validationMessage: "Something went wrong.",
                        nativeEvent: "blur",
                    },
                }),
            );
            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.vm.$data.validityMode).toBe("ERROR");
            expect(wrapper.vm.viewValue).toBe("original input");
            expect(wrapper.vm.$data.lastModelValue).toBe("original input");

            await wrapper.setProps({ modelValue: vModel });

            expect(wrapper.vm.viewValue).toBe(vModel);
        },
    );

    it.each`
        vModel   | expectedViewModel | formatter | failingFormatter | parser   | failingParser
        ${"qwe"} | ${"qwe"}          | ${false}  | ${false}         | ${false} | ${false}
        ${"foo"} | ${"FOO"}          | ${true}   | ${false}         | ${false} | ${false}
        ${"foo"} | ${"foo"}          | ${true}   | ${true}          | ${false} | ${false}
        ${"BAR"} | ${"BAR"}          | ${false}  | ${false}         | ${true}  | ${false}
        ${"BAR"} | ${"BAR"}          | ${false}  | ${false}         | ${true}  | ${true}
        ${"asd"} | ${"ASD"}          | ${true}   | ${false}         | ${true}  | ${false}
        ${"asd"} | ${"asd"}          | ${true}   | ${true}          | ${true}  | ${false}
        ${"asd"} | ${"asd"}          | ${true}   | ${false}         | ${true}  | ${true}
    `(
        "should set viewModel to '$expectedViewModel' when v-model is set to '$vModel' and formatter=$formatter, parser=$parser, failingFormatter=$failingFormatter, failingParser=$failingParser)",
        async ({
            vModel,
            expectedViewModel,
            formatter,
            parser,
            failingFormatter,
            failingParser,
        }) => {
            const formatterMock = failingFormatter
                ? () => undefined
                : (viewValue: string): string => viewValue.toUpperCase();
            const parserMock = failingParser
                ? () => undefined
                : (viewValue: string): string => viewValue.toLowerCase();

            const wrapper = createWrapper({
                props: {
                    formatter: formatter ? formatterMock : undefined,
                    parser: parser ? parserMock : undefined,
                    modelValue: "zxc",
                },
                attrs: {
                    id: "elementId",
                    "data-validation": true,
                },
            });

            await wrapper.setProps({ modelValue: vModel });

            expect(wrapper.vm.viewValue).toBe(expectedViewModel);
        },
    );
});

describe("html-validate", () => {
    it("should only allow f-tooltip in tooltip slot", () => {
        expect.assertions(2);
        const valid = /* HTML */ `<f-tooltip
            screen-reader-text="lorem ipsum"
        />`;
        const invalid = /* HTML */ `<div />`;
        const markup = (child: string): string => /* HTML */ `
            <f-text-field v-validation.maxLength>
                Label
                <template #tooltip> ${child} </template>
            </f-text-field>
        `;
        expect(markup(valid)).toHTMLValidate();
        expect(markup(invalid)).not.toHTMLValidate({
            ruleId: "element-permitted-content",
            message: `<div> element is not permitted as content under slot "tooltip" (<f-text-field>)`,
        });
    });
});
