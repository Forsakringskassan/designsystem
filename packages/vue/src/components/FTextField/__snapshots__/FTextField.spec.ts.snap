// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot when label and description slots are not specified 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="fkui-vue-element-0001"
    >
      
      
      <span>
        My defaultText
      </span>
      
      
      
      
      <span
        class="label__description"
      >
        <span
          class="sr-only"
        >
          My descriptionScreenReaderText
        </span>
        <span>
          My descriptionText
        </span>
      </span>
      <span
        class="label__description label__description--discrete"
      >
        <span
          class="sr-only"
        >
          My discreteDescriptionScreenReaderText
        </span>
        <span>
          My discreteDescriptionText
        </span>
      </span>
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="fkui-vue-element-0001"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when label and description slots are specified 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="fkui-vue-element-0001"
    >
      
      
      Custom default slot
      
      
      
      
      Custom description slot
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="fkui-vue-element-0001"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when validityMode is ERROR and isValid is false 1`] = `
<div
  class="text-field text-field--error"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="elementId"
    >
      
      
      Label
      
      
      
      
      <span
        class="label__message label__message--error"
      >
        <f-icon-stub
          class="label__icon--left"
          library="f"
          name="error"
        />
        
        
        
        Something went wrong.
        
        
        
      </span>
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="elementId"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when validityMode is INITIAL and isValid is false 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="elementId"
    >
      
      
      Label
      
      
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="elementId"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when validityMode is INITIAL and isValid is true 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="elementId"
    >
      
      
      Label
      
      
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="elementId"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when validityMode is VALID and isValid is true 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="elementId"
    >
      
      
      Label
      
      
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="elementId"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot with label and input 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="fkui-vue-element-0001"
    >
      
      
      Label
      
      
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="fkui-vue-element-0001"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot with label, error message and input 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="fkui-vue-element-0001"
    >
      
      
      Label
      
      
      
      
      <span
        class="label__message label__message--error"
      >
        <f-icon-stub
          class="label__icon--left"
          library="f"
          name="error"
        />
        
        
        ERROR_MESSAGE
        
        
      </span>
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="fkui-vue-element-0001"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot with label, tooltip, description, error message and input 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <div
      class=""
    >
      <div>
        <label
          class="label"
          for="fkui-vue-element-0001"
        >
          
          
          Label
          
          
        </label>
      </div>
      
      
      TOOLTIP
      
      
      <label
        class="label sr-separator"
        for="fkui-vue-element-0001"
      >
        
        
        DESCRIPTION
        
        
        <span
          class="label__message label__message--error"
        >
          <f-icon-stub
            class="label__icon--left"
            library="f"
            name="error"
          />
          
          
          ERROR_MESSAGE
          
          
        </span>
      </label>
    </div>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        class="text-field__input"
        id="fkui-vue-element-0001"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;
